### DISCUSSION TITLE
Additional Subscriptions info required?

### DISCUSSION THREAD

**Discussion by HilalLko** [imdhemy/laravel-in-app-purchases#21] [Q&A] [answered]
@imdhemy 
How to get subscriptions amount which is deducted when subscriptions renew as well updated order Id for the same? 
This is required to maintain transaction logs in our database (or server).

For Example :
New Order Id generated,
Amount Charged etc.

**Comment by imdhemy** [comment, marked as answer]
Hi @hilal-hipster,
Thank you for creating this topic.
Getting the subscription info is related to the context and the subscription provider. 

## From a purchase event:
```php
$subscription = $event->getSubscription()->getProviderRepresentation();
```
The `getProviderRepresentation()` method returns either `SubscriptionPurchase ` object or `ReceiptResponse ` object based on the provider. The first object represents the `Resource: SubscriptionPurchase` from [Google](https://developers.google.com/android-publisher/api-ref/rest/v3/purchases.subscriptions#SubscriptionPurchase), and the later object represents the `responseBody` from [App Store](https://developer.apple.com/documentation/appstorereceipts/responsebody).

Each key has a getter method, for example, `kind` has a getter method `getKind()`, and so on.

## Using The Subscription Facade:
Check the following sections in the main documentation
- [Google Play Subscriptions](https://github.com/imdhemy/laravel-in-app-purchases#google-play-subscriptions)
- [App Store Subscriptions](https://github.com/imdhemy/laravel-in-app-purchases#app-store-subscriptions)

## Example

```php
 /** @var Imdhemy\GooglePlay\Subscriptions\SubscriptionPurchase $subscription */
$subscription = $event->getSubscription()->getProviderRepresentation();

$orderId = $subscription->getOrderId();
$priceAmountMicros = $subscription->getPriceAmountMicros();
$priceCurrencyCode = $subscription->getPriceCurrencyCode();
```

Still not clear enough, tell me which key from which provider ([Google Play](https://developers.google.com/android-publisher/api-ref/rest/v3/purchases.subscriptions#resource:-subscriptionpurchase), [App Store](https://developer.apple.com/documentation/appstorereceipts/responsebody)) you need to retrieve and I'll provide an example showing how to get it.

**Reply by HilalLko** [reply]
Thanks for the response

but I did it using this way for Google Play

```php
        //Getting Notitifcation Object (as shown by you)
        $notification = $event->getServerNotification();
        //Getting Subscription Object from Notitifcation Object
        $subscription = $notification->getSubscription();
        //Getting Item id from from Subscription Object        
        $itemId = $notification->getSubscription()->getItemId();        
        //Getting Subscription Unique identifier
        $uniqueIdentifier = $subscription->getUniqueIdentifier();                
        //Getting Subscriptions Updated expiry Time
        $expirationTime = $subscription->getExpiryTime();
        //This is what i changed another call to Google Play subscription in order to get new Order ID
        $checkSubscription = Subscription::googlePlay()->id($itemId)->token($uniqueIdentifier)->get();
        $orderId = $checkSubscription->getOrderId();
```
as well as to get Price Macros & Price Currency Code I am utilizing the same $checkSubscription object
like

```php
$checkSubscription->getPriceCurrencyCode();
$checkSubscription->getPriceAmountMicros();
```

**Comment by HilalLko** [comment]
I have a question App Store provides getPriceAmountMicros() or not?

**Reply by imdhemy** [reply]
I'm sorry, can not remember now the full response structure of the App Store, but you can check the [documentation](https://developer.apple.com/documentation/appstorereceipts/responsebody).

**Reply by HilalLko** [reply]
Sample AppStore SubscriptionRenew example?

**Reply by HilalLko** [reply]
> Sample AppStore SubscriptionRenew example?

@imdhemy Brother any update for this, as right now am always getting the same unique identifier from AppStore in each RTDN
```php
$uniqueIdentifier = $subscription->getUniqueIdentifier();
```

**Comment by divyame1** [comment]
Hi @imdhemy 
```php
public function handle(SubscriptionExpired $event)
{
   $notification = $event->getServerNotification();
   $subscription = $notification->getSubscription();
   $uniqueIdentifier = $subscription->getSubscriptionIdentifier();
   $ids = $subscription->getSubscriptionId(); 
   Log::info(sprintf("Google Play PurchasesExp, version: %s", 1));
   Log::info(sprintf("Google Play Test NotificationExp, version: %s", $ids));
    //  Log::info(sprintf("Google Play Test Notification, version Exp: %s", $uniqueIdentifier));
}   
```

```php
public function handle(SubscriptionCanceled $event)
{
   $notification = $event->getServerNotification();
   $subscription = $notification->getSubscription();
   $uniqueIdentifier = $subscription->getSubscriptionIdentifier();
   $ids = $subscription->getSubscriptionId(); 
   Log::info(sprintf("Google Play Purchasescan, version: %s", 1));
   Log::info(sprintf("Google Play Test Notificationcan, version: %s", $ids));
   Log::info(sprintf("Google Play Test Notification, version can: %s", $uniqueIdentifier));
        
}
```    
 
- These getSubscriptionId , getSubscriptionIdentifier not working in cancel and expired subscription. ?  
- How subscription expired will work.?
- while purchasing subscription getting only subscription id , uniqueIdentifier , Can we get email id as well?
- uniqueIdentifier using this how we can check this uniqueIdentifier related to this user?
- uniqueIdentifier will be same for purchase , cancel , expired , renew etc ?
- 
 Please help
Thanks

**Reply by imdhemy** [reply]
Hi @divyame1!

Correct me if I'm wrong! You're trying to connect the subscription data with the user entity on your DB, aren't you?

In such a situation, feel free to follow the steps described in this discussion. #24 and yes! the unique identifier which is the purchase token in case of a Google play purchase will be the same for cancel, expiration, renew, etc...

**Reply by batusenyuz** [reply]
Hello, for App Store and play Store getPriceAmountMicros() and getPriceCurrencyCode methods return undefined? I need to save the subscription price and currency information in my database.

**Comment by batusenyuz** [comment]
Hello, for App Store and play Store getPriceAmountMicros() and getPriceCurrencyCode methods return undefined? I need to save the subscription price and currency information in my database.
